Reworked GPU driver installations for x86 systems

Fourdee opened the pull request:
Reworked GPU driver installations for x86 systems. 

This provides better control of GPU drivers for DietPi, adding support for Vulkan and resolves issues with Steam Proton. Users will be prompted to select a driver package during DietPi-Update.: https://github.com/MichaIng/DietPi/issues/6262

Testing:
- x86 testing only. All these changes are disabled for other Arch's. However, ARM test to check for any conflicts during patch should be done.
- :u6307: DietPi-Update >> Intel + Steam (Sonic Mania via Proton 7.0-6)
- :u6307:DietPi-Update >> Nvidia + Steam (Sonic Mania via Proton 7.0-6)
- :u6307:DietPi-Update  >> Custom
- :u6307:DietPi-Update >> None
- :u6307: DietPi-Config >> Manual test of all options.

ToDo:
AMD test.
Docs to enable proton (disabled by default):
- Load Steam
- Click View
- Settings
- SteamPlay
- ```Enable Steam Play for all other titles```

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update dietpi-config

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update dietpi-software

Fourdee committed:
Update dietpi-config

Fourdee committed:
Update patch_file

Fourdee committed:
Update dietpi-config

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update dietpi-software

Fourdee committed:
Update dietpi.txt

Fourdee committed:
Update patch_file

Fourdee committed:
Update dietpi-software

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update dietpi-config

Fourdee committed:
Update dietpi-software

Fourdee committed:
Update dietpi-config

Fourdee committed:
Update patch_file

Fourdee committed:
Update patch_file

Fourdee committed:
Update patches

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update dietpi-config

Fourdee committed:
Update dietpi-software

Fourdee committed:
Update patches

Fourdee committed:
Update dietpi.txt

Fourdee committed:
Update dietpi-software

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update dietpi-software

Fourdee committed:
Update dietpi-config

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update dietpi-software

Fourdee committed:
Update dietpi-config

Fourdee committed:
Update patches

Fourdee committed:
Update dietpi-software

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update dietpi-config

Fourdee committed:
Update dietpi.txt

Fourdee committed:
Update patches

Fourdee committed:
Update dietpi-config

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update dietpi-set_hardware

Fourdee committed:
Update patches

Fourdee committed:
Update patches

Fourdee committed:
Update CHANGELOG.txt

MichaIng commented on dietpi.txt:
```diff
@@ -161,6 +161,13 @@ CONFIG_CPU_MIN_FREQ=Disabled
+# - Intel
+# - Nvidia
+# - AMD
+CONFIG_GPU_DRIVER='None'
```
We should make clear here that it is relevant only for x86_64 PC systems when using gaming or video software (has not really an effect on common desktop usage, right?).

For simplicity, shall we just expect lower-case characters everywhere? In the `dietpi-config` menu where it looks nicer, we can  `${G_WHIP_RETURNED_VALUE,,}`. EDIT: Ah already done.

Fourdee replied:
> We should make clear here that it is relevant only for x86_64 PC systems when using gaming or video software

Also Chromium (Video decode and surface rendering) and any software that will use HW acelleration on the driver stack.

I would personally leave it as simply GPU driver. If needed,  add "Adds support for HW accell, for applications that support OpenGL/GLES/Vulkan and other API's etc".

> (has not really an effect on common desktop usage, right?).

Should take load of CPU (eg: moving windows etc). X11 will render them in OpenGL on GPU HW I believe?

Fourdee replied:
Could mention x86 only. But left it out as you mentioned ARM external support. Could be done in future?

MichaIng replied:
> Also Chromium (Video decode and surface rendering)

Jep, everything that does video playback I mean. OpenGL and GLES are installed with X11 in any case (hard package dependency), but indeed Chromium generally supports VA-API as well, AFAIR and at least for Nvidia and AMD the firmware and DDX will have an effect on desktop rendering. Ah and `nvidia-driver` extends OpenGL as well. Okay indeed just makes sense in general if any GUI application is used 👍.

Then I suggest:
```
# - x86_64 PC only!
# - Adds support for GUI/video hardware acceleration, OpenGL/GLES, Vulkan and VA-API
```
When we implement ARM support, we can remove the first line.

MichaIng commented on dietpi/dietpi-config:
```diff
@@ -534,65 +534,68 @@ A long (or insufficiently manufactured) cable may required a higher boost settin
-					G_AG_CHECK_INSTALL_PREREQ firmware-amd-graphics xserver-xorg-video-amdgpu libgl1-mesa-dri
-				fi
+				gpu_new=${G_WHIP_RETURNED_VALUE,,}
 			fi
```
`$G_WHIP_RETURNED_VALUE` is assured to be set. When hitting "Cancel" or interrupt/terminate the menu, one falls back to the parent display menu immediately.
```suggestion
			local gpu_new=${G_WHIP_RETURNED_VALUE,,}
```

MichaIng commented on dietpi/func/dietpi-set_hardware:
```diff
@@ -2259,6 +2259,96 @@ _EOF_
+		#Current GPU
+		local gpu_current=$(sed -n '/^[[:blank:]]*CONFIG_GPU_DRIVER=/{s/^[^=]*=//p;q}' /boot/dietpi.txt)
+		gpu_current=${gpu_current,,}
+		INPUT_DEVICE_VALUE=${INPUT_DEVICE_VALUE,,}
```
Already lower-case, see above 🙂, but fallback is nice to have:
```suggestion
		[[ $INPUT_DEVICE_VALUE ]] || INPUT_DEVICE_VALUE='custom'
```

Fourdee replied:
> Already lower-case, see above slightly_smiling_face, but fallback is nice to have:

Yes. Covers user use of ```dietpi-set_hardware dedicatedgpu ITEM``` and future code.

MichaIng replied:
User use is forced lower-case as well. When I find time, I want to remove the `INPUT_DEVICE_VALUE=${INPUT_DEVICE_VALUE,,}` at the top of the script indeed and require explicit cases. In this case, also here `INPUT_DEVICE_VALUE=${INPUT_DEVICE_VALUE,,}` would be removed and instead we pass it in lower case to the script in the first place (we do already in `dietpi-config`), and fail if cases are wrong. See my other comment to make them lower-care in `dietpi.txt` already.

Fourdee replied:
@MichaIng 
Just an idea. You could store the available option values for each function in dietpi-set_hardware, in an array.

Then in DietPi-_CONFIG_. You could use ```dietpi-set_hardware dedicatedgpu list```. Which would return an array, this could then be dumped directly into Whiptail menus and only offers user available options.
This would remove the need to manually type each option twice, both in ```dietpi-set_hardware``` and ```dietpi-config```.

eg:
```dietpi-set_hardware [function] list```
would return available options:
```
none
intel
nvidia
amd
custom
```

Maybe if I get time, i can do an example? But i'll finish this PR first :smile:

MichaIng replied:
> Then in DietPi-Software. You could use

You mean in `dietpi-config`? On the other hand, since we add description in the menus and often want to show the options in beautiful case-sensitive way, we anyway have to fully define the array there. So I think after all it does not make much difference regarding amount of code/repitition. Also another dedicated script call adds a significant delay on slower SBCs.*

But as you mentioned "DietPi-Software": We could actually prompt the menu `/boot/dietpi/dietpi-config 2` during Steam install, as we do on next DietPi update. At least if the current selection is still "none".

\* I'm actually planning to move parts of `dietpi-config` into dedicated scripts with CLI included, like `dietpi-vpn`, `dietpi-ddns`, `dietpi-services`, `dietpi-autostart`. E.g. adding `dietpi-network`, `dietpi-audio` (with JustBoom script included), `dietpi-display` etc. I like code/topics which belong together in a single script with menu, CLI and backend functions combined. That also eliminates the duplicates. Of course `dietpi-config` will still show all these options, but when entering e.g. "Audio Options", the related dedicated script is called. But hard to find time for such larger reworks, so this will happen one by one with delay. Currently working on larger `dietpi-services` rework: service startups at boot will be done by systemd the common way and `dietpi-services` will do only the stop/restart on installs/updates/backups and such (and when called manually of course).

Fourdee replied:
> You mean in dietpi-config? 

Yeah I did, my bad :smile: 

> But as you mentioned "DietPi-Software": We could actually prompt the menu /boot/dietpi/dietpi-config 2 during Steam install, as we do on next DietPi update. At least if the current selection is still "none".

Not a bad idea :+1: . I'd probably move it to X11 install. Prompt there if no driver set?

Fourdee replied:
>  I'm actually planning to move parts of dietpi-config into dedicated scripts with CLI included

Sounds awesome, but yeah I can imagine the rewrites and work needed for this. Massive task that one :+1:

MichaIng commented on dietpi/func/dietpi-set_hardware:
```diff
@@ -2259,6 +2259,96 @@ _EOF_
+	
+		#Current GPU
+		local gpu_current=$(sed -n '/^[[:blank:]]*CONFIG_GPU_DRIVER=/{s/^[^=]*=//p;q}' /boot/dietpi.txt)
+		gpu_current=${gpu_current,,}
```
I first though about doing a fallback to custom/noop when calling this script, but actually its easier to do it here:
```suggestion
		[[ $gpu_current ]] && gpu_current=${gpu_current,,} || gpu_current='custom'
```

MichaIng commented on dietpi/func/dietpi-set_hardware:
```diff
@@ -2259,6 +2259,96 @@ _EOF_
+				G_AGP i965-va-driver xserver-xorg-video-intel
+				G_AGP nvidia-driver nvidia-driver-libs:* libnvidia-glvkspirv:* nvidia-xconfig libgl1-nvidia-glx nvidia-settings nvidia-vulkan-common nvidia-vulkan-icd:*
+				G_AGP firmware-amd-graphics xserver-xorg-video-amdgpu
+				G_AGA
```
To pass the wildcard through and avoid globbing, it needs to be quoted. `G_AGP` checks internally against the list of installed packages which one to purge, so there the asterisk should be expanded. And for performance reasons, I'd merge all purges into a single call.

And: `G_AGP` implies `G_AGA` in the meantime, so this does not need to be called afterwards 😉:
```suggestion
				G_AGP 'mesa-vulkan-drivers:*' i965-va-driver xserver-xorg-video-intel nvidia-driver 'nvidia-driver-libs:*' 'libnvidia-glvkspirv:*' nvidia-xconfig libgl1-nvidia-glx nvidia-settings nvidia-vulkan-common 'nvidia-vulkan-icd:*' firmware-amd-graphics xserver-xorg-video-amdgpu
```
Is there actually a chance that other package depend on any of these and might be unintentionally purged? We did not take care of this before, but to be failsafe we could also do:
```suggestion
				apt-mark auto $(dpkg --get-selections 'mesa-vulkan-drivers:*' i965-va-driver xserver-xorg-video-intel nvidia-driver 'nvidia-driver-libs:*' 'libnvidia-glvkspirv:*' nvidia-xconfig libgl1-nvidia-glx nvidia-settings nvidia-vulkan-common 'nvidia-vulkan-icd:*' firmware-amd-graphics xserver-xorg-video-amdgpu 2> /dev/null | mawk '{print $1}') 2> /dev/null && G_AGA
```

Fourdee replied:
> And for performance reasons, I'd merge all purges into a single call.

Yep, I wanted to keep the driver list per HW vendor on a dedicated line. Was easier to work with in code and see what package is for which device.

Maybe using your ```mark auto```, but adding? 
```
String='AMd packages'
String+=' Intel packages'
String+=' Nvidia Packages'
mark auto $String
```

MichaIng replied:
> Maybe using your `mark auto`, but adding?

Jep, that will also work, same as we do below for adding kernel headers with an array (best practice shell coding, so all entries are individually quoted):
```sh
array=('AMD1' 'AMD2')
array+=('Intel1' 'Intel2')
array+=('Nvidia1' 'Nvidia2')
apt-mark auto "${array[@]}"
```
Same can be applied to the other installs to always have a single `G_AGI` or better `G_AG_CHECK_INSTALL_PREREQ` call as well. The latter implies `G_AGUP` if required and does nothing if all packages are installed already.

Generally an alternative for are escaped line breaks:
```sh
apt-mark auto 1 2 3\
	4 5 6\
	7 8 9
```

MichaIng commented on .update/patches:
```diff
@@ -1193,6 +1193,17 @@ Patch_8_16()
+	then	
+	
+		G_WHIP_MSG "DietPi GPU driver installations have been reworked for x86 platforms.\nThis resolves issues with Vulkan, Steam and improves overall driver package installations.\n\nOn the next page you will be asked to install a GPU driver for this device. Once completed, select 'Cancel' from the menu to resume DietPi-Updates.\n\nNB: If you have a custom GPU driver you have manually installed on th ...
+		/boot/dietpi/dietpi-config 2
```
We should only show this menu if any GPU driver or Steam is installed already. Would be confusing for users with VMs or headless/server systems. Just not sure how to obtain this info best. Steam software install state and checking for one of the main packages we checked before in the `dietpi-config` menu: `nvidia-driver` `i965-va-driver` `xserver-xorg-video-amdgpu`?

Fourdee replied:
Yep. We should only need to check for those installed drivers via dpkg as you mentioned. Then we can offer this prompt.
Steam will apply i386 extras during GPU driver install automatically, if Steam is installed.

MichaIng replied:
Jep, also `nvidia-driver-libs:i386` is already installed if Steam was installed as of the Nvidia device check we did before during install:
```sh
[[ -e '/sys/module/nvidia/version' ]] && nvidia=('nvidia-driver-libs:i386')
```

Fourdee replied:
https://github.com/MichaIng/DietPi/pull/6273/commits/149017fa04efd2fd4f61c93a2756e08cd3bde615

Fourdee committed:
https://github.com/MichaIng/DietPi/pull/6273#discussion_r1148430200

> We should only show this menu if any GPU driver or Steam is installed already. Would be confusing for users with VMs or headless/server systems.

Fourdee committed:
Syntax

Fourdee committed:
https://github.com/MichaIng/DietPi/pull/6273#discussion_r1148595606

Fourdee committed:
https://github.com/MichaIng/DietPi/pull/6273#discussion_r1148427355

Fourdee committed:
https://github.com/MichaIng/DietPi/pull/6273#discussion_r1148427699

Fourdee committed:
https://github.com/MichaIng/DietPi/pull/6273#pullrequestreview-1357895507

Fourdee committed:
https://github.com/MichaIng/DietPi/pull/6273#pullrequestreview-1357896421

MichaIng commented on dietpi/func/dietpi-set_hardware:
```diff
@@ -2259,6 +2259,84 @@ _EOF_
+				local uninstall_array=('mesa-vulkan-drivers:*') #Mesa Vulkan
+				uninstall_array+=('i965-va-driver' 'xserver-xorg-video-intel') #Intel
+				uninstall_array+=('nvidia-driver' 'nvidia-driver-libs:*' 'libnvidia-glvkspirv:*' 'nvidia-xconfig' 'libgl1-nvidia-glx' 'nvidia-settings' 'nvidia-vulkan-common' 'nvidia-vulkan-icd:*') #Nvidia
+				uninstall_array+=('firmware-amd-graphics' 'xserver-xorg-video-amdgpu') #AMD
```
Let's not purge everything first what is installed later again. This means unnecessary downloads and file writes and in case config removals. As the array is split already, we can add only those parts which are not to be installed:
```suggestion
				local uninstall_array=()
				[[ $INPUT_DEVICE_VALUE == 'intel' ]] || uninstall_array+=('i965-va-driver' 'xserver-xorg-video-intel')
				[[ $INPUT_DEVICE_VALUE == 'nvidia' ]] || uninstall_array+=('nvidia-driver' 'nvidia-driver-libs:*' 'libnvidia-glvkspirv:*' 'nvidia-settings' 'nvidia-vulkan-common' 'nvidia-vulkan-icd:*' 'nvidia-xconfig' 'libgl1-nvidia-glx')
				[[ $INPUT_DEVICE_VALUE == 'amd' ]] || uninstall_array+=('firmware-amd-graphics' 'xserver-xorg-video-amdgpu')
				[[ $INPUT_DEVICE_VALUE =~ ^intel|amd$ ]] || uninstall_array+=('mesa-vulkan-drivers:*')
```
Not sure about the Mesa Vulkan driver. If "custom" is selected, one might want (to keep) this generic package?

Fourdee replied:
> Not sure about the Mesa Vulkan driver. If "custom" is selected, one might want (to keep) this generic package?

If custom is selected. No uninstalls or installs take place. https://github.com/Fourdee/DietPi/blob/dev/dietpi/func/dietpi-set_hardware#L2282-L2286

MichaIng commented on dietpi/func/dietpi-set_hardware:
```diff
@@ -2259,6 +2259,84 @@ _EOF_
+			if [[ $INPUT_DEVICE_VALUE == 'intel' ]]
+			then
+
+				install_array+=('i965-va-driver' 'xserver-xorg-video-intel' 'libgl1-mesa-dri' 'mesa-vulkan-drivers')
```
Let's not install the legacy Intel DDX. All cards from 2007 and above work better with the default modesetting DDX according to package description.
```suggestion
				install_array+=('i965-va-driver' 'libgl1-mesa-dri' 'mesa-vulkan-drivers')
```

Fourdee committed:
https://github.com/MichaIng/DietPi/pull/6273#discussion_r1150997947

Fourdee committed:
https://github.com/MichaIng/DietPi/pull/6273#issuecomment-1486976364

Fourdee committed:
https://github.com/MichaIng/DietPi/pull/6273#discussion_r1150996506

Fourdee committed:
Remove trailing spaces

Fourdee committed:
Resolve trailing spaces

Fourdee committed:
Resolve trailing spaces

Fourdee committed:
Resolve trailing spaces

Fourdee committed:
gpudriver naming

Fourdee committed:
gpudriver naming

Fourdee committed:
gpudriver naming

Fourdee committed:
SC2053

Fourdee committed:
SC2053

Fourdee committed:
Prompt GPU driver install

Offer GPU driver install at end of Software installs

Fourdee committed:
Fix

Lower case and unquote

Fourdee committed:
SC2086

Fourdee committed:
SC2086

Fourdee committed:
Tweak

Only offer GPU driver install when Xserver is installed

MichaIng committed:
v8.16

- General | Add missing execute permissions

MichaIng committed:
v8.16

- DietPi-Patches | Check for i965-va-driver package as Intel driver stack indicator, since we do not install the legacy xserver-xorg-video-intel anymore. Fix and align menu a little, and show it on first package match instead of checking for all of them.

MichaIng committed:
v8.16

- DietPi-Patches | Offer GPU driver selection as well if Steam is installed
- DietPi-Set_hardware | GPU driver: Mute error messages when marking packages to be uninstalled and only do autoremoval if there are any packages to be uninstalled
